source/game_sa/Entity/Vehicle/Vehicle.cpp

```diff
@@ -4277,18 +4282,102 @@ void CVehicle::DoHeadLightBeam(eVehicleDummy dummyId, CMatrix& matrix, bool arg2
 }
 
 // 0x6E1440
-void CVehicle::DoHeadLightReflectionSingle(CMatrix& matrix, uint8 lightId) {
-    ((void(__thiscall*)(CVehicle*, CMatrix&, uint8))0x6E1440)(this, matrix, lightId);
+void CVehicle::DoHeadLightReflectionSingle(CMatrix& matrix, bool isRight) {
+    const auto  rwObject  = CModelInfo::ms_modelInfoPtrs[m_nModelIndex]->AsVehicleModelInfoPtr()->m_pRwObject;
+    const auto  type      = isRight ? *reinterpret_cast<const float*>(&rwObject->type) : -*reinterpret_cast<const float*>(&rwObject->type);
+    const auto  parent    = *reinterpret_cast<const float*>(&rwObject->parent);
+    const auto& position  = m_matrix->GetPosition();
+    CVector     shadowPos = position;
+
+    float upLength    = sqrt(matrix.GetUp().x * 2.f + matrix.GetUp().y * 2.f);
+    float rightLength = sqrt(matrix.GetRight().x * 2.f + matrix.GetRight().y * 2.f);
+
+    float invertXb = matrix.GetUp().x / upLength;
+    float _y1      = matrix.GetUp().y / upLength;
+
+    float matrixb = matrix.GetRight().x / rightLength;
+    float v23     = matrix.GetRight().y / rightLength;
+
+    float v14 = (IsBike() || IsQuad()) ? 1.25f : fabs(type) * 4.0f;
+    float v15 = v14 * 2.0f + 1.0f + parent;
+
+    shadowPos.x += matrixb * type + invertXb * v15;
+    shadowPos.y += v23 * type + _y1 * v15;
+    shadowPos.z += 2.0f;
+
+    const auto x1 = 2.0f * v14 * invertXb;
+    const auto y1 = 2.0f * v14 * _y1;
+    const auto x2 = v14 * _y1;
+    const auto y2 = -v14 * invertXb;
+    CShadows::StoreCarLightShadow(
+        this,
+        reinterpret_cast<int32>(&m_matrix) + 2, // fix reversed please
+        gpShadowHeadLightsTex2,
+        shadowPos,
+        x1, y1,
+        x2, y2,
+        45u, 45u, 45u,
+        7.0f
+    );
 }
```

Comment:
Untested, please test
```suggestion
void CVehicle::DoHeadLightReflectionSingle(CMatrix& lightMat, uint8 lightId) {
    const auto& lightDummyPos = GetDummyPositionObjSpace(DUMMY_LIGHT_FRONT_MAIN);
    const auto  lightFwd2D    = CVector2D{lightMat.GetForward()}.Normalized();
    const auto  lightRight2D  = CVector2D{lightMat.GetRight()}.Normalized();
    const auto  size          = IsSubBike() || GetModelID() == MODEL_QUAD
        ? 1.25
        : std::fabs(lightDummyPos.x) * 4.f;
    const auto posOS = CVector2D{ // 2D position of the light, in object space
        lightDummyPos.x,
        lightDummyPos.y + size * 2.f + 1.f
    };
    const auto shdwFront = lightFwd2D * (size * 2.f);
    const auto shdwSide  = (lightFwd2D * size).GetPerpRight();
    CShadows::StoreCarLightShadow(
        this,
        reinterpret_cast<int32>(&this->m_matrix) + 2,
        gpShadowHeadLightsTex2,
        GetPosition() + CVector{posOS.Dot(lightRight2D), posOS.Dot(lightFwd2D), 2.f}, // Basically a 2D matrix transform here
        shdwFront.x, shdwFront.y,
        shdwSide.x, shdwSide.y,
        45, 45, 45,
        7.f
    );
}
```